Plots/Contours/NCL_eof_1_1.py

```diff
@@ -0,0 +1,290 @@
+
+deg2rad = 4. * atan(1.) / 180.
+clat = SLP['lat']
+clat = sqrt(cos(deg2rad * clat))
+print(clat)
+
+###############################################################################
+# Multiply SLP by weights.
+#
+# Xarray uses the supplied coordinate information to apply latitude-based
+# weights to all longitudes and timesteps automatically.
+
+wSLP = SLP
+wSLP['slp'] = clat * SLP['slp']
+
+# For now, metadata for slp must be copied over explicitly; it is not preserved by binary operators like multiplication.
+wSLP['slp'].attrs = ds['slp'].attrs
+wSLP['slp'].attrs['long_name'] = 'Wgt: ' + wSLP['slp'].attrs['long_name']
+
+###############################################################################
+# Subset data to the North Atlantic region.
+
+xw = wSLP.sel(lat=slice(latS, latN), lon=slice(lonL, lonR))
+
+print('\n\nxw:\n\n')
+print(xw.slp)
+
+###############################################################################
+# Compute the EOFs.
+
+eof = eofunc(xw["slp"], neof, time_dim=1, meta=True)
+
+print('\n\neof:\n\n')
+print(eof)
+
+eof_ts = eofunc_ts(xw["slp"], eof, time_dim=1, meta=True)
+
+print('\n\neof_ts:\n\n')
+print(eof_ts)
+
+
+###############################################################################
+# Normalize time series: Sum spatial weights over the area used.
+nLon = len(xw['lon'])
+clat_subset = clat.where((clat.lat >= latS) & (clat.lat <= latN), drop=True)
+weightTotal = clat_subset.sum() * nLon
+eof_ts = eof_ts / weightTotal
+
+print('\n\neof_ts normalized:\n\n')
+print(eof_ts)
+
+
+###############################################################################
+# Define a utility function for creating a contour plot.
+
+def make_contour_plot(ax, dataset):
+
+    map_extent = [lonL, lonR, latS, latN]
+    ax.set_extent(map_extent, crs=ccrs.PlateCarree())
+
+    # Use geocat.viz.util convenience function to set axes tick values
+    gvutil.set_axes_limits_and_ticks(ax, xlim=None, ylim=None, xticks=[-60, -30, 0, 30], yticks=[40, 60, 80])
+
+    # Use geocat.viz.util convenience function to add minor and major tick lines
+    gvutil.add_major_minor_ticks(ax, x_minor_per_major=3, y_minor_per_major=4, labelsize=10)
+
+    lat = dataset['lat']
+    lon = dataset['lon']
+    values = dataset.data
+
+    cmap = gvcmaps.BlWhRe
+
+    # Specify contour levels
+    v = np.linspace(-0.08, 0.08, 9, endpoint=True)
+
+    # The function contourf() produces fill colors, and contour() calculates contour label locations.
+    cplot = ax.contourf(lon, lat, values, levels=v, cmap=cmap, extend="both")
+    p = ax.contour(lon, lat, values, levels=v, linewidths=0.0)
+    ax.clabel(p, fontsize=8, fmt="%0.2f", colors="k")
+
+    ax.coastlines(linewidth=0.5)
+    gvutil.add_lat_lon_ticklabels(ax)
+    return cplot, ax
+
+
+###############################################################################
+# Draw a contour plot for each EOF.
+
+fig, axs = plt.subplots(neof, 1,
+                        constrained_layout=True,  # "magic"
```

Comment:
```suggestion
```

This should fix it.

Reply:
Hi Deepak, when I remove 'constrained_layout=True',  it shrinks the plot axes and gives too much of the figure space to the colorbar.    It seems wrong, like matplotlib could not possibly have these configuration settings as defaults.  Have you encountered this before?   I have looked at the other examples in the NCL gallery with multiple map subplots, and they all seem to have 'constrained_layout=True'.

Reply:
You could try `plt.tight_layout()` instead

Otherwise I would tweak figure size till constrained_layout does OK. `constrained_layout` uses a linear constraint solver; sometimes you don't find a solution and it issues that warning. https://matplotlib.org/tutorials/intermediate/constrainedlayout_guide.html

Reply:
https://matplotlib.org/tutorials/intermediate/constrainedlayout_guide.html#debugging

Reply:
I have tried debugging, and the warning about constrained_layout will not go away even if the plot consists of three simple boxes:
![test](https://user-images.githubusercontent.com/10638475/77676208-9de1a180-6f53-11ea-9bc8-3ffd3bcc95d0.png)

It looks to be an outstanding bug when using constrained_layout and cartopy together.

@erogluorhan   I have run out of time working on this example.   Please let me know if you are able to take over changes to this pull request.  Thank you!

Reply:
BTW I tried using both matplotlib 3.1.3 and 3.2.1.   Both versions had this problem with the warning message.   However, the plot (after a call to plt.savefig()) looks fine despite the warning message.

Reply:
@bonnland Thanks a lot for this great effort! Yes, let me take over this code and see if I can put anything beyond. I will get into every single comment of yours and let you know about my findings, if any.